Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-134.ca-central-1.compute.internal
Sha: 5df1b40a8b5d020070f3044bf4b66851d6aad442
Number of Error(s): 19
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
CMake Error at /home/centos/smoketest/PR-14842/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
make[2]: *** [CMakeFiles/generate-boost-regex.dir/build.make:90: generate-boost-regex-prefix/src/generate-boost-regex-stamp/generate-boost-regex-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:937: CMakeFiles/generate-boost-regex.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
CMake Error at /home/centos/smoketest/PR-14842/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
  Each download failed!
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
         status_code: 22
         status_string: HTTP response code said error
         log:
         --- LOG BEGIN ---
           Trying 52.12.180.214...
  TCP_NODELAY set
  Connected to dl.bintray.com (52.12.180.214) port 443 (#0)
  ALPN, offering http/1.1
  successfully set certificate verify locations:
    CAfile: /etc/pki/tls/certs/ca-bundle.crt
    CApath: none
  [5 bytes data]
  TLSv1.3 (OUT), TLS handshake, Client hello (1):
  [512 bytes data]
  [5 bytes data]
  TLSv1.3 (IN), TLS handshake, Server hello (2):
  [102 bytes data]
  NPN, negotiated HTTP1.1
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Certificate (11):
  [2765 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server key exchange (12):
  [333 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server finished (14):
  [4 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
  [70 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
  [1 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Next protocol (67):
  [36 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Finished (20):
  [16 bytes data]
  [5 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Finished (20):
  [16 bytes data]
  SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
  ALPN, server did not agree to a protocol
  Server certificate:
   subject: CN=*.bintray.com
   start date: Sep 26 00:00:00 2019 GMT
   expire date: Nov  9 12:00:00 2021 GMT
   subjectAltName: host dl.bintray.com matched cert\s *.bintray.com
   issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 2018
   SSL 
 ... ( comment is too long, 1188 lines, 100608 bytes, truncated) ... 
[Build error:](https://gist.github.com/40387f51a01663732450821b83cb1fe9)
Build log [Build log](https://gist.github.com/HPCCSmoketest/40387f51a01663732450821b83cb1fe9#file-result-21-04-12-14-10-20-log)

```

--- a/dali/daliadmin/daadmin.cpp
+++ b/dali/daliadmin/daadmin.cpp
@@ -59,6 +59,9 @@
 
 #include "daadmin.hpp"
 
+namespace daadmin
+{
+
 static unsigned daliConnectTimeoutMs = 5000;
 
 static bool noninteractive=false;
@@ -139,7 +142,7 @@ void setDaliConnectTimeoutMs(unsigned timeoutMs)
 
 //=============================================================================
 
-void _export_(const char *path,const char *dst,bool safe)
+void exportToFile(const char *path,const char *filename,bool safe)
 {
     StringBuffer xpath;
     Owned<IRemoteConnection> conn = connectXPathOrFile(path,safe,xpath);
@@ -148,52 +151,48 @@ void _export_(const char *path,const char *dst,bool safe)
         return;
     }
     Owned<IPropertyTree> root = conn->getRoot();
-    Owned<IFile> f = createIFile(dst);
+    Owned<IFile> f = createIFile(filename);
     Owned<IFileIO> io = f->open(IFOcreate);
     Owned<IFileIOStream> fstream = createBufferedIOStream(io);
     toXML(root, *fstream);          // formatted (default)
-    OUTLOG("Branch %s saved in '%s'",xpath.str(),dst);
+    OUTLOG("Branch %s saved in '%s'",xpath.str(),filename);
     conn->close();
+}
+
+//=============================================================================
+
+bool exportToXML(const char *path,StringBuffer &out,bool safe)
+{
+    StringBuffer xpath;
+    Owned<IRemoteConnection> conn = connectXPathOrFile(path,safe,xpath);
+    if (!conn) {
+        out.appendf("Could not connect to %s",path);
+        return false;
+    }
+    Owned<IPropertyTree> root = conn->getRoot();
+    toXML(root, out);
+    conn->close();
+    return true;
 }
 
 //==========================================================================================================
 
-void import(const char *path,const char *src,bool add)
+bool doImport(const char *path,const char *head,const char *tail,const char *xml,bool add,StringBuffer &out)
 {
-    Owned<IFile> iFile = createIFile(src);
-    Owned<IFileIO> iFileIO = iFile->open(IFOread);
-    if (!iFileIO)
-    {
-        UERRLOG("Could not open to %s",src);
-        return;
-    }
-    size32_t sz = (size32_t)iFile->size();
-    StringBuffer xml;
-    iFileIO->read(0, sz, xml.reserve(sz));
-    Owned<IPropertyTree> branch = createPTreeFromXMLString(xml.str());
-    StringBuffer head;
-    StringBuffer tmp;
-    const char *tail=splitpath(path,head,tmp);
-    if (!add) {
-        Owned<IRemoteConnection> bconn = querySDS().connect(remLeading(path),myProcessSession(),RTM_LOCK_READ|RTM_SUB, daliConnectTimeoutMs);
-        if (bconn) {
-            Owned<IPropertyTree> broot = bconn->getRoot();
-            StringBuffer bakname;
-            Owned<IFileIO> io = createUniqueFile(NULL, tail, "bak", bakname);
-            OUTLOG("Saving backup of %s to %s",path,bakname.str());
-            Owned<IFileIOStream> fstream = createBufferedIOStream(io);
-            toXML(broot, *fstream);         // formatted (default)
-        }
-    }
-    Owned<IRemoteConnection> conn = querySDS().connect(head.str(),myProcessSession(),0, daliConnectTimeoutMs);
+    Owned<IPropertyTree> branch = createPTreeFromXMLString(xml);
+    Owned<IRemoteConnection> conn = querySDS().connect(head,myProcessSession(),0, daliConnectTimeoutMs);
     if (!conn) {
-        UERRLOG("Could not connect to %s",path);
-        return;
+        out.appendf("Could not connect to %s",path);
+        return false;
     }
     StringAttr newtail; // must be declared outside the following if
     Owned<IPropertyTree> root = conn->getRoot();
     if (!add) {
         Owned<IPropertyTree> child = root->getPropTree(tail);
+        if (!child) {
+            out.appendf("Invalid path: %s",path);
+            return false;
+        }
         root->removeTree(child);
 
         //If replacing a qualified branch then remove the qualifiers before calling addProp
@@ -210,42 +209,82 @@ void import(const char *path,const char *src,bool add)
         oldEnvironment.setown(createPTreeFromIPT(conn->queryRoot()));
     root->addPropTree(tail,LINK(branch));
     conn->commit();
-    OUTLOG("Branch %s loaded from '%s'",path,src);
     conn->close();
     if (*path=='/')
         path++;
     if (strcmp(path,"Environment")==0) {
-        OUTLOG("Refreshing cluster groups from Environment");
+        out.appendf("Refreshing cluster groups from Environment.");
         StringBuffer response;
         initClusterGroups(false, response, oldEnvironment);
         if (response.length())
-            PROGLOG("updating Environment via import path=%s : %s", path, response.str());
+            out.appendf(" Updating Environment via import path=%s : %s.", path, response.str());
     }
+    return true;
+}
+
+bool importFromXML(const char *path,const char *xml,bool add,StringBuffer &out)
+{
+    StringBuffer head,tmp;
+    const char *tail=splitpath(path,head,tmp);
+    return doImport(path,head,tail,xml,add,out);
+}
+
+bool importFromFile(const char *path,const char *filename,bool add,StringBuffer &out)
+{
+    Owned<IFile> iFile = createIFile(filename);
+    Owned<IFileIO> iFileIO = iFile->open(IFOread);
+    if (!iFileIO)
+    {
+        out.appendf("Could not open to %s",filename);
+        return false;
+    }
+    size32_t sz = (size32_t)iFile->size();
+    StringBuffer xml;
+    iFileIO->read(0, sz, xml.reserve(sz));
+    StringBuffer head;
+    StringBuffer tmp;
+    const char *tail=splitpath(path,head,tmp);
+    if (!add) {
+        Owned<IRemoteConnection> bconn = querySDS().connect(remLeading(path),myProcessSession(),RTM_LOCK_READ|RTM_SUB, daliConnectTimeoutMs);
+        if (bconn) {
+            Owned<IPropertyTree> broot = bconn->getRoot();
+            StringBuffer bakname;
+            Owned<IFileIO> io = createUniqueFile(NULL, tail, "bak", bakname);
+            out.appendf("Saving backup of %s to %s",path,bakname.str());
+            Owned<IFileIOStream> fstream = createBufferedIOStream(io);
+            toXML(broot, *fstream);         // formatted (default)
+        }
+    }
+
+    bool ret = doImport(path, head, tail, xml, add, out);
+    if (ret)
+        out.appendf(" Branch %s loaded from '%s'",path,filename);
+    return ret;
 }
 
 //=============================================================================
 
 
-void _delete_(const char *path,bool backup)
+bool erase(const char *path,bool backup,StringBuffer &out)
 {
     StringBuffer head;
     StringBuffer tmp;
     const char *tail=splitpath(path,head,tmp);
     Owned<IRemoteConnection> conn = querySDS().connect(head.str(),myProcessSession(),RTM_LOCK_WRITE, daliConnectTimeoutMs);
     if (!conn) {
-        UERRLOG("Could not connect to %s",path);
-        return;
+        out.appendf("Could not connect to %s",path);
+        return false;
     }
     Owned<IPropertyTree> root = conn->getRoot();
     Owned<IPropertyTree> child = root->getPropTree(tail);
     if (!child) {
-        UERRLOG("Couldn't find %s/%s",head.str(),tail);
-        return;
+        out.appendf("Couldn't find %s/%s",head.str(),tail);
+        return false;
     }
     if (backup) {
         StringBuffer bakname;
         Owned<IFileIO> io = createUniqueFile(NULL,"daliadmin", "bak", bakname);
-        OUTLOG("Saving backup of %s/%s to %s",head.str(),tail,bakname.str());
+        out.appendf("Saving backup of %s/%s to %s",head.str(),tail,bakname.str());
         Owned<IFileIOStream> fstream = createBufferedIOStream(io);
         toXML(child, *fstream);         // formatted (default)
     }
@@ -254,6 +293,7 @@ void _delete_(const char *path,bool backup)
     root.clear();
     conn->commit();
     conn->close();
+    return true;
 }
 
 //=============================================================================
@@ -3252,3 +3292,5 @@ void removeOrphanedGlobalVariables(bool dryrun, bool reconstruct)
         PROGLOG("%sChecked: [%u / %u] - deletes: %u, auxDeletes: %u, persists: %u. [Avg ms: %2.2f]", dryrun?"DRYRUN:":"", checked, total, deletes, auxDeletes, existingPersists, static_cast<unsigned>(cycle_to_microsec(avgCycles))/1000.0);
     }
 }
+
+} // namespace daadmin

--- a/dali/daliadmin/daadmin.hpp
+++ b/dali/daliadmin/daadmin.hpp
@@ -23,13 +23,18 @@
 
 #include "dautils.hpp"
 
+namespace daadmin
+{
+
 extern DALIADMIN_API void setDaliConnectTimeoutMs(unsigned timeoutMs);
 extern DALIADMIN_API void xmlSize(const char *filename, double pc);
 extern DALIADMIN_API void translateToXpath(const char *logicalfile, DfsXmlBranchKind tailType = DXB_File);
 
-extern DALIADMIN_API void _export_(const char *path, const char *dst, bool safe = false);
-extern DALIADMIN_API void import(const char *path, const char *src, bool add);
-extern DALIADMIN_API void _delete_(const char *path, bool backup);
+extern DALIADMIN_API void exportToFile(const char *path, const char *filename, bool safe = false);
+extern DALIADMIN_API bool exportToXML(const char *path, StringBuffer &out, bool safe = false);
+extern DALIADMIN_API bool importFromFile(const char *path, const char *filename, bool add, StringBuffer &out);
+extern DALIADMIN_API bool importFromXML(const char *path, const char *xml, bool add, StringBuffer &out);
+extern DALIADMIN_API bool erase(const char *path, bool backup,StringBuffer &out);
 extern DALIADMIN_API StringBuffer &setValue(const char *path, const char *val, StringBuffer &oldVal);
 extern DALIADMIN_API void getValue(const char *path, StringBuffer& out);
 extern DALIADMIN_API void bget(const char *path, const char *outfn);
@@ -89,3 +94,5 @@ extern DALIADMIN_API void daliping(const char *dalis, unsigned connecttime, unsi
 
 extern DALIADMIN_API void validateStore(bool fix, bool deleteFiles, bool verbose);
 extern DALIADMIN_API void removeOrphanedGlobalVariables(bool dryrun, bool reconstruct);
+
+} // namespace daadmin

--- a/dali/daliadmin/daliadmin.cpp
+++ b/dali/daliadmin/daliadmin.cpp
@@ -20,7 +20,17 @@
 #include "rmtfile.hpp"
 #include "daadmin.hpp"
 
+using namespace daadmin;
+
 #define DEFAULT_DALICONNECT_TIMEOUT 5 // seconds
+
+void doLog(bool noError, StringBuffer &out)
+{
+    if (noError)
+        PROGLOG("%s", out.str());
+    else
+        UERRLOG("%s", out.str());
+}
 
 void usage(const char *exe)
 {
@@ -253,20 +263,20 @@ int main(int argc, const char* argv[])
                     setDaliConnectTimeoutMs(1000 * props->getPropInt("timeout", DEFAULT_DALICONNECT_TIMEOUT));
                     if (strieq(cmd,"export")) {
                         CHECKPARAMS(2,2);
-                        _export_(params.item(1),params.item(2));
+                        exportToFile(params.item(1),params.item(2));
                     }
                     else if (strieq(cmd,"import")) {
                         CHECKPARAMS(2,2);
-                        import(params.item(1),params.item(2),false);
+                        doLog(importFromFile(params.item(1),params.item(2),false,out),out);
                     }
                     else if (strieq(cmd,"importadd")) {
                         CHECKPARAMS(2,2);
-                        import(params.item(1),params.item(2),true);
+                        doLog(importFromFile(params.item(1),params.item(2),true,out),out);
                     }
                     else if (strieq(cmd,"delete")) {
                         CHECKPARAMS(1,2);
                         bool backup = np<2 || !strieq("nobackup", params.item(2));
-                        _delete_(params.item(1),backup);
+                        doLog(erase(params.item(1),backup,out),out);
                     }
                     else if (strieq(cmd,"set")) {
                         CHECKPARAMS(2,2);

--- a/esp/scm/ws_dali.ecm
+++ b/esp/scm/ws_dali.ecm
@@ -23,7 +23,7 @@ ESPrequest [nil_remove] SetValueRequest
     string Value;
 };
 
-ESPresponse [exceptions_inline, nil_remove] ValueResponse
+ESPresponse [exceptions_inline, nil_remove] ResultResponse
 {
     string Result;
 };
@@ -33,11 +33,25 @@ ESPrequest [nil_remove] GetValueRequest
     string Path;
 };
 
-ESPservice [auth_feature("NONE"), //This declares that the method logic handles feature level authorization
-    version("1.01"), default_client_version("1.01"), exceptions_inline("./smc_xslt/exceptions.xslt")] WSDali
+ESPrequest [nil_remove] ImportRequest
 {
-    ESPmethod [min_ver("1.01")] SetValue(SetValueRequest, ValueResponse);
-    ESPmethod [min_ver("1.01")] GetValue(GetValueRequest, ValueResponse);
+    string XML;
+    string Path;
+    bool Add;
+};
+
+ESPrequest [nil_remove] DeleteRequest
+{
+    string Path;
+};
+
+ESPservice [auth_feature("NONE"), //This declares that the method logic handles feature level authorization
+    version("1.02"), default_client_version("1.02"), exceptions_inline("./smc_xslt/exceptions.xslt")] WSDali
+{
+    ESPmethod [min_ver("1.01")] SetValue(SetValueRequest, ResultResponse);
+    ESPmethod [min_ver("1.01")] GetValue(GetValueRequest, ResultResponse);
+    ESPmethod [min_ver("1.02")] Import(ImportRequest, ResultResponse);
+    ESPmethod [min_ver("1.02")] Delete(DeleteRequest, ResultResponse);
 };
 
 SCMexportdef(WSDali);

--- a/esp/services/ws_dali/ws_daliservice.cpp
+++ b/esp/services/ws_dali/ws_daliservice.cpp
@@ -25,6 +25,8 @@
 #include "dasds.hpp"
 #include "daadmin.hpp"
 
+using namespace daadmin;
+
 #define REQPATH_EXPORTSDSDATA "/WSDali/Export"
 
 const char* daliFolder = "tempdalifiles" PATHSEPSTR;
@@ -107,7 +109,7 @@ void CWSDaliEx::checkAccess(IEspContext& context)
         throw makeStringException(ECLWATCH_CANNOT_CONNECT_DALI, "Dali detached.");
 }
 
-bool CWSDaliEx::onSetValue(IEspContext& context, IEspSetValueRequest& req, IEspValueResponse& resp)
+bool CWSDaliEx::onSetValue(IEspContext& context, IEspSetValueRequest& req, IEspResultResponse& resp)
 {
     try
     {
@@ -135,7 +137,7 @@ bool CWSDaliEx::onSetValue(IEspContext& context, IEspSetValueRequest& req, IEspV
     return true;
 }
 
-bool CWSDaliEx::onGetValue(IEspContext& context, IEspGetValueRequest& req, IEspValueResponse& resp)
+bool CWSDaliEx::onGetValue(IEspContext& context, IEspGetValueRequest& req, IEspResultResponse& resp)
 {
     try
     {
@@ -148,6 +150,52 @@ bool CWSDaliEx::onGetValue(IEspContext& context, IEspGetValueRequest& req, IEspV
         StringBuffer value;
         getValue(path, value);
         resp.setResult(value);
+    }
+    catch(IException* e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWSDaliEx::onImport(IEspContext& context, IEspImportRequest& req, IEspResultResponse& resp)
+{
+    try
+    {
+        checkAccess(context);
+
+        const char* xml = req.getXML();
+        const char* path = req.getPath();
+        if (isEmptyString(xml))
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "Data XML not specified.");
+        if (isEmptyString(path))
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "Data path not specified.");
+
+        StringBuffer result;
+        if (importFromXML(path, xml, req.getAdd(), result))
+            result.appendf(" Branch %s loaded.", path);
+        resp.setResult(result);
+    }
+    catch(IException* e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWSDaliEx::onDelete(IEspContext& context, IEspDeleteRequest& req, IEspResultResponse& resp)
+{
+    try
+    {
+        checkAccess(context);
+
+        const char* path = req.getPath();
+        if (isEmptyString(path))
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "Data path not specified.");
+
+        StringBuffer result;
+        erase(path, false, result);
+        resp.setResult(result);
     }
     catch(IException* e)
     {

--- a/esp/services/ws_dali/ws_daliservice.hpp
+++ b/esp/services/ws_dali/ws_daliservice.hpp
@@ -48,8 +48,10 @@ public:
     }
 
     virtual void init(IPropertyTree* cfg, const char* process, const char* service) override;
-    virtual bool onSetValue(IEspContext& context, IEspSetValueRequest& req, IEspValueResponse& resp) override;
-    virtual bool onGetValue(IEspContext& context, IEspGetValueRequest& req, IEspValueResponse& resp) override;
+    virtual bool onSetValue(IEspContext& context, IEspSetValueRequest& req, IEspResultResponse& resp) override;
+    virtual bool onGetValue(IEspContext& context, IEspGetValueRequest& req, IEspResultResponse& resp) override;
+    virtual bool onImport(IEspContext& context, IEspImportRequest& req, IEspResultResponse& resp) override;
+    virtual bool onDelete(IEspContext& context, IEspDeleteRequest& req, IEspResultResponse& resp) override;
 };
 
 class CWSDaliSoapBindingEx : public CWSDaliSoapBinding

